src/hotspot/share/opto/loopnode.cpp

```diff
@@ -1720,47 +1638,279 @@ void PhaseIdealLoop::check_counted_loop_shape(IdealLoopTree* loop, Node* x, Basi
+  assert(stride.compute_non_zero_stride_con(exit_test.mask(), bt), "illegal condition");
 
-  PhiNode* phi = loop_iv_phi(xphi, phi_incr, x);
+  assert(exit_test.mask() != BoolTest::ne, "unexpected condition");
+  assert(iv_incr.phi_incr() == nullptr, "bad loop shape");
+  assert(exit_test.cmp()->in(1) == iv_incr.incr(), "bad exit test shape");
 
-  assert(phi != nullptr && phi->in(LoopNode::LoopBackControl) == incr, "No phi");
+  // Safepoint on backedge not supported
+  assert(x->in(LoopNode::LoopBackControl)->Opcode() != Op_SafePoint, "no safepoint on backedge");
+}
+#endif
 
-  jlong stride_con = stride->get_integer_as_long(bt);
+void PhaseIdealLoop::LoopExitTest::build() {
+  _is_valid = false;
 
-  assert(condition_stride_ok(mask, stride_con), "illegal condition");
+  const Node* iftrue = _back_control;
+  uint iftrue_op = iftrue->Opcode();
+  Node* iff = iftrue->in(0);
+  BoolNode* test = iff->in(1)->as_Bool();
+  _mask = test->_test._test;
+  _cl_prob = iff->as_If()->_prob;
+  if (iftrue_op == Op_IfFalse) {
+    _mask = BoolTest(_mask).negate();
+    _cl_prob = 1.0f - _cl_prob;
+  }
+  // Get backedge compare
+  _cmp = test->in(1);
+  if (!_cmp->is_Cmp()) {
+    return;
+  }
 
-  assert(mask != BoolTest::ne, "unexpected condition");
-  assert(phi_incr == nullptr, "bad loop shape");
-  assert(cmp->in(1) == incr, "bad exit test shape");
+  // Find the trip-counter increment & limit.  Limit must be loop invariant.
+  _incr  = _cmp->in(1);
+  _limit = _cmp->in(2);
 
-  // Safepoint on backedge not supported
-  assert(x->in(LoopNode::LoopBackControl)->Opcode() != Op_SafePoint, "no safepoint on backedge");
+  // ---------
+  // need 'loop()' test to tell if limit is loop invariant
+  // ---------
+
+  if (!_phase->is_member(_loop, _phase->get_ctrl(_incr))) { // Swapped trip counter and limit?
+    swap(_incr, _limit);   // Then reverse order into the CmpI
+    _mask = BoolTest(_mask).commute(); // And commute the exit test
+  }
+  if (_phase->is_member(_loop, _phase->get_ctrl(_limit))) { // Limit must be loop-invariant
+    return;
+  }
+  if (!_phase->is_member(_loop, _phase->get_ctrl(_incr))) { // Trip counter must be loop-variant
+    return;
+  }
+
+  _is_valid = true;
+}
+
+// Canonicalize the loop condition if it is 'ne'.
+bool PhaseIdealLoop::LoopExitTest::canonicalize_mask(jlong stride_con) {
+  if (_mask != BoolTest::ne) {
+    return false;
+  }
+
+  assert(stride_con == 1 || stride_con == -1, "simple increment only - checked in CountedLoopConverter");
+
+  if (stride_con == 1) {
+    // 'ne' can be replaced with 'lt' only when init < limit.
+    // This is ensured by the inserted predicate in CountedLoopConverter
+    _mask = BoolTest::lt;
+    return true;
+  }
+
+  if (stride_con == -1) {
+    // 'ne' can be replaced with 'gt' only when init > limit.
+    // This is ensured by the inserted predicate in CountedLoopConverter.
+    _mask = BoolTest::gt;
+    return true;
+  }
+
+  // Should not reach
+  return false;
```

Comment:
With the assert above for `stride_con` being 1 or -1, do we need this `if` here? Or are you concerned about when this does not hold in product? But since we are not using the return value anyways, I don't think it makes a difference and you could just drop the `if`.